### PR TITLE
refactor(io): create a style guide for error messages, align error messages to the style guide

### PR DESCRIPTION
**Pull request by irbull**
Creates an initial style guide for user-facing error messages. The guide is based on the style guide from V8.

### CODE REVIEW

File: .github/CONTRIBUTING.md
Comment (kt3k) [inline .github/CONTRIBUTING.md:196]: I wonder if this rule is reasonable. If the value is a string, using quotes around it doesn't seem like a problem to me.

How about not having this rule?

File: .github/CONTRIBUTING.md
Comment (kt3k) [inline .github/CONTRIBUTING.md:229]: The example part seems having an error?
```suggestion
Good: Cannot parse input x: value is empty
```

File: .github/CONTRIBUTING.md
Comment (irbull) [reply .github/CONTRIBUTING.md:196]: IMHO I like the quotes too, so +1 from me. I just used what was suggested on the other PR and moved it here. I can update the PR if you'd like, however I'm at the lake with extremely bad internet this week, so I don't know if I'll be able to push the change. I'm back to civilization on Sunday, so I can do it then.

File: .github/CONTRIBUTING.md
Comment (irbull) [reply .github/CONTRIBUTING.md:229]: Opps. From the original suggestion it was "Should the message use a colon or comma when providing additional information". Do you prefer colon, comma, or either?

**Comment by irbull** [comment]
I'd be happy to take a first pass at the std to see if it conforms to these rules. I'd start by looking at places an `error` is raised. Let me know if you think this would help?

File: .github/CONTRIBUTING.md
Comment (iuioiua) [reply .github/CONTRIBUTING.md:229]: We prefer colon.

Comment (irbull) [reply .github/CONTRIBUTING.md]: I've updated the guide to mention that colon is preferred.

File: .github/CONTRIBUTING.md
Comment (irbull) [reply .github/CONTRIBUTING.md]: I've updated the style guide to indicate that quotes should be used for values.

**Comment by irbull** [comment]
I've updated the `CONTRIBUTING.md` with the suggestions that @kt3k had. I also took a pass at the `io/` folder in a second commit on this PR. If you like this approach, I can do some of the other folders too. Or if you would prefer I can file them in different PRs.

File: io/buffer.ts
Comment (kt3k) [inline io/buffer.ts:356]:
```suggestion
        `The buffer cannot be grown beyond the maximum size of "${MAX_SIZE}"`,
```

File: .github/CONTRIBUTING.md
Comment (kt3k) [inline .github/CONTRIBUTING.md:202]: Thanks for updating. But I think this rule needs more consideration. For example, using quotes makes sense when the value is a string, but it probably makes more sense to not use quotes when the value is a number. I guess there are edge cases where it's difficult to decide which pattern to use (nested object, etc).

For now, I'd suggest removing this rule (until we come up with a good way to formulate this)

```suggestion
```

**Comment by kt3k** [comment]
Most other part looks good to me

File: io/read_range.ts
Comment (timreichen) [inline io/read_range.ts:49]: According to the guide line shouldn't this be something like `Byte range start cannot be larger than end`?

File: io/read_range.ts
Comment (timreichen) [inline io/read_range.ts:100]: According to the guide line shouldn't this be something like `Byte range start cannot be larger than end`?

File: .github/CONTRIBUTING.md
Comment (jsejcksn) [reply .github/CONTRIBUTING.md:202]: I think quoting strings is sensible… and I've often used direct JSON serializations (e.g. ``const message = "…found `${JSON.stringify(str)}`…";``) in my own error messages to get quotes for free (and representations that can be copy+pasted back into some JS source). The downside to that is when it's desirable to print literals which are replaced by escaped sequences in the output (e.g. line breaks).

**Comment by jsejcksn** [comment]
Linking https://github.com/denoland/std/issues/5574

File: .github/CONTRIBUTING.md
Comment (irbull) [reply .github/CONTRIBUTING.md:202]: This is why I attempted to updated the error messages in a folder, I think this gives us some good concrete examples to talk about and see how they _feel_.

I'm happy to remove this rule, or we could also use this:

3. Message should use quotes for values for strings:

```
Bad: Cannot parse input hello, world
Good: Cannot parse input "hello, world"
Good: Attempting to grow buffer by 10 bytes, which would exceed the maximum size of 100 bytes
```

What do you think, do you want to keep this rule for Strings?

File: io/buffer.ts
Comment (irbull) [reply io/buffer.ts:356]: Yikes, sorry about that. I'll update the PR.

**Comment by irbull** [comment]
> Linking:
> 
> * [Normalize error message grammar #5574](https://github.com/denoland/std/issues/5574)

Thanks for linking! I've added a table to the issue that lists all the packages. I'll take a stab at the other ones once we're happy with the style and I return to some more stable internet.

File: .github/CONTRIBUTING.md
Comment (kt3k) [reply .github/CONTRIBUTING.md]: > 3. Message should use quotes for values for strings:
> 
> ```
> Bad: Cannot parse input hello, world
> Good: Cannot parse input "hello, world"
> Good: Attempting to grow buffer by 10 bytes, which would exceed the maximum size of 100 bytes
> ```

This looks good to me

**Comment by irbull** [comment]
While updating a few more error messages, I think having either comma `,` or colon `:` for additional information makes sense. In some cases, it flows much better with commas. For example:

```
Messagepack decode came across an invalid type for a key of a map, keys must be a number or a string
```
vs
```
Messagepack decode came across an invalid type for a key of a map: keys must be a number or a string
```

**Comment by irbull** [comment]
I've filed another PR with a few more folders (toml, msgpack & crypto). I'm not sure if you want all the commits here, or if it's easier to review / merge if we split them up.

**Comment by kt3k** [comment]
To maintainers (@iuioiua, @bartlomieju): What commit tag should we be using for these style changes of error messages? `fix`, `docs`, or `refactor`?

**Comment by iuioiua** [comment]
The docs changes in `.github/CONTRIBUTING.md` should be it's own PR with the enforcing of these rules in other PRs.

**Comment by iuioiua** [comment]
Historically, the changing of these error messages have been refactors.

**Comment by irbull** [comment]
> The docs changes in `.github/CONTRIBUTING.md` should be it's own PR with the enforcing of these rules in other PRs.
That's fair. I created separate commits so that should be easy. I wanted to include a few changes so we could see how these error messages look. I'll separate the initial commit into its own PR.

> Historically, the changing of these error messages have been refactors.
I can update the commit message to use `refactor` instead of `fix`. The first commit should be `doc` I guess, as it's the documentation change.

**Comment by bartlomieju** [comment]
> To maintainers (@iuioiua, @bartlomieju): What commit tag should we be using for these style changes of error messages? `fix`, `docs`, or `refactor`?

I'm fine with `refactor`

**Comment by bartlomieju** [comment]
Ooops, sorry, closed by accident 😓

**Comment by irbull** [comment]
Would you like me to pull out the changes to the error messages in the `io` folder and leave this PR simply for creating the style guide? I typically `--amend` commits and make each commit (and message) the way I like it and then FF merge instead of squashing all the commits in a PR when merging. This way I can see the way each commit will look before merging, it's a bit more work, but it ensures that the commit & commit message are reviewed together.

But if you would prefer, I can pull out the `io` commit from this PR, create a separate PR for that and then we can just focus on the style guide here? Let me know.

**Review by kt3k — APPROVED** [review]
Let's just land this. Thanks for your contribution!

**Pull request merged**
